File: app/src/main/java/ru/systempla/weatherapp/mvp/view/ui/fragment/WeatherDataFragment.java
package ru.systempla.weatherapp.mvp.view.ui.fragment;

import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.text.InputType;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.View;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.RelativeLayout;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.widget.PopupMenu;
import androidx.core.view.GravityCompat;
import androidx.drawerlayout.widget.DrawerLayout;

import java.util.Date;
import java.util.Locale;

import butterknife.BindDrawable;
import butterknife.BindString;
import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import butterknife.Unbinder;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;
import moxy.MvpAppCompatFragment;
import moxy.presenter.InjectPresenter;
import moxy.presenter.ProvidePresenter;
import ru.systempla.weatherapp.R;
import ru.systempla.weatherapp.mvp.App;
import ru.systempla.weatherapp.mvp.presenter.WeatherDataPresenter;
import ru.systempla.weatherapp.mvp.view.WeatherDataView;
import timber.log.Timber;

@SuppressWarnings({"unused","ConstantConditions"})
public class WeatherDataFragment extends MvpAppCompatFragment implements WeatherDataView {

    private Unbinder unbinder;
    private DrawerLayout drawer;

    public static WeatherDataFragment newInstance(){
        return new WeatherDataFragment();
    }

    @InjectPresenter
    WeatherDataPresenter presenter;

    @BindString(R.string.language)
    String language;

    @BindString(R.string.wind_speed)
    String speedUnit;

    @BindString(R.string.pressure_unit)
    String pressureUnit;

    @BindView(R.id.loading)
    RelativeLayout loadingRelativeLayout;

    @BindView(R.id.optionHitBoxExtender)
    View popupMenuButton;

    @BindView(R.id.n_city_label)
    TextView cityLabel;

    @BindView(R.id.n_humidity_value)
    TextView humidityValue;

    @BindView(R.id.n_pressure_value)
    TextView pressureValue;

    @BindView(R.id.n_uv_value)
    TextView uvValue;

    @BindView(R.id.n_weather_desc_value)
    TextView weatherDescValue;

    @BindView(R.id.n_weather_image)
    ImageView weatherImage;

    @BindView(R.id.n_wind_speed_value)
    TextView windSpeedValue;

    @BindView(R.id.n_temperature_value)
    TextView temperatureValue;

    @BindView(R.id.drawer_button)
    ImageView drawerButton;

    @BindDrawable(R.drawable.ic_sun)
    Drawable iconSun;

    @BindDrawable(R.drawable.ic_moon)
    Drawable iconMoon;

    @BindDrawable(R.drawable.ic_snow)
    Drawable iconSnow;

    @BindDrawable(R.drawable.ic_fog)
    Drawable iconFog;

    @BindDrawable(R.drawable.ic_rain)
    Drawable iconRain;

    @BindDrawable(R.drawable.ic_cloud)
    Drawable iconCloud;

    @BindDrawable(R.drawable.ic_thunderstorm)
    Drawable iconThunderstorm;

    @BindDrawable(R.drawable.ic_drizzle)
    Drawable iconDrizzle;

    @SuppressWarnings("WeakerAccess")
    @OnClick(R.id.optionHitBoxExtender)
    public void showMenu(View v){
        showPopupMenu(v);
    }

    @SuppressWarnings("WeakerAccess")
    @OnClick(R.id.drawer_button)
    public void showDrawer(){drawer.openDrawer(GravityCompat.START);}

    private void showPopupMenu(View v){
        PopupMenu popupMenu = new PopupMenu(this.getContext(), v);
        popupMenu.inflate(R.menu.location_change_menu);
        popupMenu.setOnMenuItemClickListener(item -> {
            switch (item.getItemId()) {
                case R.id.change_city:
                    showInputDialog();
                    return true;
                case R.id.change_to_gps:
                    presenter.setSetting("gps");
                    presenter.loadAccordingToSettings();
                    return true;
                default:
                    return false;
            }
        });
        popupMenu.show();
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);
    }

    @Override
    public void onResume() {
        super.onResume();
        presenter.setLanguage(language);
        presenter.checkSettings();
        presenter.loadAccordingToSettings();
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable android.view.ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_weather_data, container, false);
        unbinder = ButterKnife.bind(this, view);
        App.getInstance().getAppComponent().inject(this);
        drawer = getActivity().findViewById(R.id.main_drawer_layout);
        return view;
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        unbinder.unbind();
    }

    @Override
    public void onCreateOptionsMenu(@NonNull Menu menu, @NonNull MenuInflater inflater) {
        inflater.inflate(R.menu.location_change_menu, menu);
        super.onCreateOptionsMenu(menu, inflater);
    }

    private void showInputDialog() {
        AlertDialog.Builder builder = new AlertDialog.Builder(this.getContext());
        builder.setTitle(R.string.change_city);

        final EditText input = new EditText(this.getContext());
        input.setInputType(InputType.TYPE_CLASS_TEXT);
        builder.setView(input);
        builder.setPositiveButton("OK", (dialog, which) -> {
            presenter.setSetting(input.getText().toString());
            presenter.loadAccordingToSettings();
        });
        builder.show();
    }

    @SuppressWarnings("WeakerAccess")
    @ProvidePresenter
    public WeatherDataPresenter providePresenter() {
        WeatherDataPresenter presenter = new WeatherDataPresenter(AndroidSchedulers.mainThread(), Schedulers.io());
        App.getInstance().getAppComponent().inject(presenter);
        return presenter;
    }

    @Override
    public void showMessage(String text) {
        Toast.makeText(this.getContext(), text, Toast.LENGTH_SHORT).show();
    }

    @Override
    public void setCityName(String name) {
        cityLabel.setText(name);
    }

    @Override
    public void setWeatherDescription(String description) {
        weatherDescValue.setText(description.toLowerCase());
    }

    @Override
    public void setUVIndex(float uvIndex) {
        String currentTextText = String.format(Locale.getDefault(), "%.2f", uvIndex);
        uvValue.setText(currentTextText);
    }

    @Override
    public void setPressure(float pressure) {
        if (language.equals("en")) {
            pressureValue.setText(String.format("%s " + pressureUnit, pressure));
        }
        if (language.equals("ru")) {
            double newPressure = pressure * 0.75006375541921;
            pressureValue.setText(String.format("%.0f " + pressureUnit, newPressure));

        }

    }

    @Override
    public void setHumidity(float humidity) {
        humidityValue.setText(String.format("%s %%", humidity));
    }

    @Override
    public void setWindSpeed(double speed) {
        windSpeedValue.setText(String.format("%s "+speedUnit, speed));
    }

    @Override
    public void setWeatherIcon(int actualId, long sunrise, long sunset) {
        int id = actualId / 100;
        Drawable icon = null;

        if (actualId == 800) {
            long currentTime = new Date().getTime();
            if (currentTime >= sunrise && currentTime < sunset) {
                icon = iconSun;
            } else {
                icon = iconMoon;
            }
        } else {
            switch (id) {
                case 2: {
                    icon = iconThunderstorm;
                    break;
                }
                case 3: {
                    icon = iconDrizzle;
                    break;
                }
                case 5: {
                    icon = iconRain;
                    break;
                }
                case 6: {
                    icon = iconSnow;
                    break;
                }
                case 7: {
                    icon = iconFog;
                    break;
                }
                case 8: {
                    icon = iconCloud;
                    break;
                }
            }
        }
        weatherImage.setImageDrawable(icon);
    }

    @Override
    public void setCurrentTemperature(float temp) {
        String currentTextText = String.format(Locale.getDefault(), "%.0f", temp);
        temperatureValue.setText(currentTextText);
    }

    @Override
    public void showLoading() {
        loadingRelativeLayout.setVisibility(View.VISIBLE);
    }

    @Override
    public void hideLoading() {
        loadingRelativeLayout.setVisibility(View.GONE);
    }
}

File: app/src/main/java/ru/systempla/weatherapp/mvp/view/ui/fragment/WeatherDataFragment.kt
package ru.systempla.weatherapp.mvp.view.ui.fragment

import android.Manifest
import android.content.DialogInterface
import android.graphics.drawable.Drawable
import android.os.Bundle
import android.text.InputType
import android.view.*
import android.widget.*
import androidx.appcompat.app.AlertDialog
import androidx.core.view.GravityCompat
import androidx.drawerlayout.widget.DrawerLayout
import butterknife.*
import com.tbruyelle.rxpermissions3.RxPermissions
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import moxy.MvpAppCompatFragment
import moxy.presenter.InjectPresenter
import moxy.presenter.ProvidePresenter
import ru.systempla.weatherapp.R
import ru.systempla.weatherapp.mvp.App
import ru.systempla.weatherapp.mvp.presenter.WeatherDataPresenter
import ru.systempla.weatherapp.mvp.view.WeatherDataView
import java.util.*


class WeatherDataFragment : MvpAppCompatFragment(), WeatherDataView {

    companion object {
        fun newInstance(): WeatherDataFragment = WeatherDataFragment()
    }

    @InjectPresenter
    lateinit var presenter: WeatherDataPresenter

    @BindView(R.id.loading)
    lateinit var loadingRelativeLayout: RelativeLayout

    @BindView(R.id.optionHitBoxExtender)
    lateinit var popupMenuButton: View

    @BindView(R.id.n_city_label)
    lateinit var cityLabel: TextView

    @BindView(R.id.n_humidity_value)
    lateinit var humidityValue: TextView

    @BindView(R.id.n_pressure_value)
    lateinit var pressureValue: TextView

    @BindView(R.id.n_uv_value)
    lateinit var uvValue: TextView

    @BindView(R.id.n_weather_desc_value)
    lateinit var weatherDescValue: TextView

    @BindView(R.id.n_weather_image)
    lateinit var weatherImage: ImageView

    @BindView(R.id.n_wind_speed_value)
    lateinit var windSpeedValue: TextView

    @BindView(R.id.n_temperature_value)
    lateinit var temperatureValue: TextView

    @BindView(R.id.drawer_button)
    lateinit var drawerButton: ImageView

    @BindString(R.string.language)
    lateinit var language: String

    @BindString(R.string.wind_speed)
    lateinit var speedUnit: String

    @BindString(R.string.pressure_unit)
    lateinit var pressureUnit: String

    @BindDrawable(R.drawable.ic_sun)
    lateinit var iconSun: Drawable

    @BindDrawable(R.drawable.ic_moon)
    lateinit var iconMoon: Drawable

    @BindDrawable(R.drawable.ic_snow)
    lateinit var iconSnow: Drawable

    @BindDrawable(R.drawable.ic_fog)
    lateinit var iconFog: Drawable

    @BindDrawable(R.drawable.ic_rain)
    lateinit var iconRain: Drawable

    @BindDrawable(R.drawable.ic_cloud)
    lateinit var iconCloud: Drawable

    @BindDrawable(R.drawable.ic_thunderstorm)
    lateinit var iconThunderstorm: Drawable

    @BindDrawable(R.drawable.ic_drizzle)
    lateinit var iconDrizzle: Drawable

    @OnClick(R.id.optionHitBoxExtender)
    fun showMenu(v: View?) {
        showPopupMenu(v!!)
    }

    @OnClick(R.id.drawer_button)
    fun showDrawer() {
        drawer.openDrawer(GravityCompat.START)
    }

    @ProvidePresenter
    fun providePresenter(): WeatherDataPresenter {
        val presenter = WeatherDataPresenter(AndroidSchedulers.mainThread(), Schedulers.io())
        App.instance.appComponent.inject(presenter)
        return presenter
    }

    private lateinit var rxPermissions: RxPermissions
    private lateinit var unbinder: Unbinder
    private lateinit var drawer: DrawerLayout


    private fun showPopupMenu(view: View) {
        val popupMenu = PopupMenu(context, view)
        popupMenu.inflate(R.menu.location_change_menu)
        popupMenu.setOnMenuItemClickListener { item: MenuItem ->
            when (item.itemId) {
                R.id.change_city -> {
                    showInputDialog()
                    return@setOnMenuItemClickListener true
                }
                R.id.change_to_gps -> {
                    presenter.setPermissionsSetting(1)
                    presenter.setCitySetting("gps")
                    presenter.loadAccordingToSettings()
                    return@setOnMenuItemClickListener true
                }
                else -> return@setOnMenuItemClickListener false
            }
        }
        popupMenu.show()
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true)
    }

    override fun onResume() {
        super.onResume()
        presenter.setLanguage(language)
        presenter.checkSettings()
        presenter.loadAccordingToSettings()
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view: View = inflater.inflate(R.layout.fragment_weather_data, container, false)
        App.instance.appComponent.inject(this)
        unbinder = ButterKnife.bind(this, view)
        rxPermissions = RxPermissions(this)
        drawer = requireActivity().findViewById(R.id.main_drawer_layout)
        return view
    }

    override fun onDestroyView() {
        super.onDestroyView()
        unbinder.unbind()
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        inflater.inflate(R.menu.location_change_menu, menu)
        super.onCreateOptionsMenu(menu, inflater)
    }

    private fun showInputDialog() {
        val builder = AlertDialog.Builder(this.requireContext())
        builder.setTitle(R.string.change_city)
        val input = EditText(this.context)
        input.inputType = InputType.TYPE_CLASS_TEXT
        builder.setView(input)
        builder.setPositiveButton("OK") { _: DialogInterface?, _: Int ->
            presenter.setCitySetting(input.text.toString())
            presenter.loadAccordingToSettings()
        }
        builder.show()
    }

    override fun checkForGPSUpdate() {
        rxPermissions.requestEachCombined(Manifest.permission.ACCESS_FINE_LOCATION,
                Manifest.permission.ACCESS_COARSE_LOCATION)
                .subscribe { permission ->
                    when {
                        permission.granted -> presenter.loadGPSData()
                        permission.shouldShowRequestPermissionRationale ->
                            showMessage("Location permission refused temporally")
                        else -> {
                            presenter.setPermissionsSetting(0)
                            showMessage("Location permission refused permanently")
                        }
                    }
                }
    }

    override fun showMessage(text: String?) {
        Toast.makeText(this.context, text, Toast.LENGTH_SHORT).show()
    }

    override fun setCityName(name: String) {
        cityLabel.text = name
    }

    override fun setWeatherDescription(description: String) {
        weatherDescValue.text = description.toLowerCase(Locale.ROOT)
    }

    override fun setUVIndex(uvIndex: Float) {
        val currentTextText = String.format(Locale.getDefault(), "%.2f", uvIndex)
        uvValue.text = currentTextText
    }

    override fun setPressure(pressure: Float) {
        if (language == "en") {
            pressureValue.text = String.format("%s $pressureUnit", pressure)
        }
        if (language == "ru") {
            val newPressure = pressure * 0.75006375541921
            pressureValue.text = String.format("%.0f $pressureUnit", newPressure)
        }
    }

    override fun setHumidity(humidity: Float) {
        humidityValue.text = String.format("%s %%", humidity)
    }

    override fun setWindSpeed(speed: Double) {
        windSpeedValue.text = String.format("%s $speedUnit", speed)
    }

    override fun setWeatherIcon(actualId: Int, sunrise: Long, sunset: Long) {
        val id = actualId / 100
        var icon: Drawable? = null
        if (actualId == 800) {
            val currentTime = Date().time
            icon = if (currentTime in sunrise until sunset) {
                iconSun
            } else {
                iconMoon
            }
        } else {
            when (id) {
                2 -> {
                    icon = iconThunderstorm
                }
                3 -> {
                    icon = iconDrizzle
                }
                5 -> {
                    icon = iconRain
                }
                6 -> {
                    icon = iconSnow
                }
                7 -> {
                    icon = iconFog
                }
                8 -> {
                    icon = iconCloud
                }
            }
        }
        weatherImage.setImageDrawable(icon)
    }

    override fun setCurrentTemperature(temp: Float) {
        val currentTextText = String.format(Locale.getDefault(), "%.0f", temp)
        temperatureValue.text = currentTextText
    }

    override fun showLoading() {
        loadingRelativeLayout.visibility = View.VISIBLE
    }

    override fun hideLoading() {
        loadingRelativeLayout.visibility = View.GONE
    }
}